### DISCUSSION TITLE
What is the correct way to create OTC/TTC with theirs tables shared

### DISCUSSION THREAD

**Discussion by TakWolf** [fonttools/fonttools#3228] [General]
I am not sure if the following codes is correct (or recommended), I only infer them through existing APIs:

```python
def create_xtf(name, is_ttf: bool):
   builder = FontBuilder(units_per_em, isTTF=is_ttf)
  # Some build codes
  return builder.font


ttc = TTCollection()
ttc.fonts.append(create_xtf('my font 1', true))
ttc.fonts.append(create_xtf('my font 2', true))
ttc.fonts.append(create_xtf('my font 3', true))
ttc.save('my-font.ttc', shareTable=True)
```

I did successfully create a `ttc` file. But it seems that the table is not shared. What did I do wrong？

**Comment by justvanrossum** [comment]
I was going to suggest this:
```python
ttc = TTCollection(shareTables=True)
```
but it seems that only has effect when reading a TTC.

**Reply by TakWolf** [reply]
Yes, I discovered this problem. Currently TTCollection().save() `shareTable` seems only effective when read `ttc`

**Reply by justvanrossum** [reply]
It seems that `ttc.save(path, shareTables=True)` should take care of if indeed. Are you sure your table data is sharable?

**Reply by TakWolf** [reply]
Yes. I have an actual project. https://github.com/TakWolf/ark-pixel-font use https://github.com/TakWolf/pixel-font-builder

It is similar to  https://github.com/adobe-fonts/source-han-sans in strategy, most of the glyph are the same, only a few glyphs have 
language flavor

https://github.com/TakWolf/pixel-font-builder/blob/1052ab0ffdcfdb28cf61ecfd6ca712c6280db23b/src/pixel_font_builder/opentype.py#L312-L316

**Reply by TakWolf** [reply]
![39a3d3ac47504910d528be11024a08b6](https://github.com/fonttools/fonttools/assets/6064962/698ff5c8-47d6-4dd2-b888-8fd6388f0d83)
![28908702b4c9f8cb0ab8fa19249d9f25](https://github.com/fonttools/fonttools/assets/6064962/fbb96034-a3ac-45b9-904e-1f4811bd0784)

**Reply by justvanrossum** [reply]
I tried to reproduce with a simple script, but it seems to correctly share identical tables:
[build_ttc.py.zip](https://github.com/fonttools/fonttools/files/12160329/build_ttc.py.zip)

**Reply by justvanrossum** [reply]
Can you post a (link to a) TTC binary that you built that isn't sharing tables?

**Reply by TakWolf** [reply]
Yes, let me do some organizing

**Reply by justvanrossum** [reply]
Btw. if even a single glyph is different, then the table containing the glyph data cannot be shared. 

In that case it may help to include the exact same glyph set in all fonts, but build different cmaps for each.

**Reply by TakWolf** [reply]
> In that case it may help to include the exact same glyph set in all fonts, but build different cmaps for each.

I overlooked this. Let me check it.

**Comment by TakWolf** [comment]
Still invalid.

[example-fonts.zip](https://github.com/fonttools/fonttools/files/12160632/example-fonts.zip)

These two font created by this script: https://github.com/TakWolf/pixel-font-builder/blob/master/examples/demo.py

https://github.com/TakWolf/pixel-font-builder/blob/master/examples/demo.py#L167C1-L171C71

I merged otf 1000 times, only with different family names.
<img width="521" alt="image" src="https://github.com/fonttools/fonttools/assets/6064962/8ee89929-c77f-4218-9ae6-3e181b3d862b">
<img width="603" alt="image" src="https://github.com/fonttools/fonttools/assets/6064962/fc77d5fd-71a2-4847-8001-a36d203530ca">

The remaining issue is that I won't decide is that should tables object are required the same memory objects?

**Reply by TakWolf** [reply]
Should the fonts use the same memory objects for the tables, only then can be shared?
I don't see that your example requires the same memory object

**Reply by justvanrossum** [reply]
The data for the table to be shared needs to be identical before it can be shared. It does not need to be the same memory object, but the have to compare equal.

I now see what's going on: the font name is included in the CFF table, too, so it is the name that makes the CFF table not sharable in your case :(

Perhaps you can try passing the same data to `setupCFF()` in all cases, and only make differences in the name table.

**Reply by punchcutter** [reply]
The way Source Han does this is by pulling the CFF table from the default Japanese font and inserting it into all the others.
